pkg/virt-controller/watch/vm_initializer.go

```diff
@@ -0,0 +1,363 @@
+	"kubevirt.io/kubevirt/pkg/controller"
+	"kubevirt.io/kubevirt/pkg/kubecli"
+	"kubevirt.io/kubevirt/pkg/log"
+)
+
+type VirtualMachineInitializer struct {
+	vmPresetInformer cache.SharedIndexInformer
+	vmInitInformer   cache.SharedIndexInformer
+	clientset        kubecli.KubevirtClient
+	queue            workqueue.RateLimitingInterface
+	store            cache.Store
+}
+
+const initializerMarking = "presets.virtualmachines.kubevirt.io"
+
+func NewVirtualMachineInitializer(vmPresetInformer cache.SharedIndexInformer, vmInitInformer cache.SharedIndexInformer, queue workqueue.RateLimitingInterface, vmInitCache cache.Store, clientset kubecli.KubevirtClient) *VirtualMachineInitializer {
+	vmi := VirtualMachineInitializer{
+		vmPresetInformer: vmPresetInformer,
+		vmInitInformer:   vmInitInformer,
+		clientset:        clientset,
+		queue:            queue,
+		store:            vmInitCache,
+	}
+	return &vmi
+}
+
+func (c *VirtualMachineInitializer) Run(threadiness int, stopCh chan struct{}) {
+	defer controller.HandlePanic()
+	defer c.queue.ShutDown()
+	log.Log.Info("Starting Virtual Machine Initializer.")
+
+	// Wait for cache sync before we start the pod controller
+	cache.WaitForCacheSync(stopCh, c.vmPresetInformer.HasSynced, c.vmInitInformer.HasSynced)
+
+	// Start the actual work
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	log.Log.Info("Stopping controller.")
+}
+
+func (c *VirtualMachineInitializer) runWorker() {
+	for c.Execute() {
+	}
+}
+
+func (c *VirtualMachineInitializer) Execute() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	err := c.execute(key.(string))
+
+	if err != nil {
+		log.Log.Reason(err).Infof("reenqueuing VM %v", key)
+		c.queue.AddRateLimited(key)
+	} else {
+		log.Log.V(4).Infof("processed VM %v", key)
+		c.queue.Forget(key)
+	}
+	return true
+}
+
+func (c *VirtualMachineInitializer) execute(key string) error {
+
+	// Fetch the latest VM state from cache
+	obj, exists, err := c.store.GetByKey(key)
+
+	if err != nil {
+		return err
+	}
+
+	// If the VM isn't in the cache, it was just deleted, so shouldn't
+	// be initialized
+	if exists {
+		var vm *kubev1.VirtualMachine
+		vm = obj.(*kubev1.VirtualMachine)
+		// only process VM's that aren't initialized by this controller yet
+		if !isInitialized(vm) {
+			return c.initializeVirtualMachine(vm)
+		}
+	}
+
+	return nil
+}
+
+func (c *VirtualMachineInitializer) initializeVirtualMachine(vm *kubev1.VirtualMachine) error {
+	// All VM's must be marked as initialized or they are held in limbo forever
+	// Collect all errors and defer returning until after the update
+	logger := log.Log
+	errors := []error{}
+
+	logger.Object(vm).Info("Initializing VirtualMachine")
+
+	allPresets, err := listPresets(c.vmPresetInformer, vm.GetNamespace())
+
+	matchingPresets := filterPresets(allPresets, vm)
+
+	if len(matchingPresets) != 0 {
+		err = applyPresets(vm, matchingPresets)
```

Comment:
the errors need to be separated from the application of the presets.

You should loop through the presets twice.
* First loop aggregates all the merge conflict errors. 
* Second loop applies the presets to the VM if no errors occurred. 

example from pod preset logic.
```
        // detect merge conflict
        err = safeToApplyPodPresetsOnPod(pod, matchingPPs)
        if err != nil {
                // conflict, ignore the error, but raise an event
                glog.Warningf("conflict occurred while applying podpresets: %s on pod: %v err: %v",
                        strings.Join(presetNames, ","), pod.GetGenerateName(), err)
                return nil
        }

        applyPodPresetsOnPod(pod, matchingPPs)
```

They are first determining if the pod preset is safe to apply, If not they record the aggregated errors.

Once it is determined they are safe to apply, they then apply the presets to the pod.

Reply:
Are you sure? Each preset is all-or-nothing so is atomic.

The only reason to loop twice is if we're taking the approach of "if any problem occurs anywhere, don't apply any presets at all". My understanding is that the approach we're taking is "if there's a problem with this preset, don't apply this one, but keep going".

Reply:
@stu-gott is right. We want to have a different behavior to regular PodPresets here, because we want to allow people to "override" a preset.
From the other perspective this means that we accept if a single field merge fails, send an event, and continue with the next preset.

Reply:
Obviously preset ordering has a profound effect with the second approach. Basically presets conflicting with each other...